### ISSUE TITLE
MPI barrier to synchronize ensembles

### ISSUE THREAD

**Issue by eirrgang** [kassonlab/gmxapi#231]
For MPI-based ensemble management, there are various ways race conditions could develop because seemingly benign assumptions can be very use-case specific. MPI barriers before and/or after simulations could mitigate some race conditions and even avoid (or at least clarify) some MPI errors (stemming from untagged mismatched MPI calls), and the barriers shouldn't introduce much cost. Still, they should probably be optional.

1. Barrier before starting simulation can make sure that no simulation launches until the MPI environment is in the expected state everywhere.
2. Barrier after simulation makes sure that no process proceeds to try to access artifacts or proceed with additional MPI calls until all simulations in the ensemble are complete.

These should probably both be optional to account for different CPU billing heuristics, but should default to "on". Situations that are not mitigated by the barriers are mismatched calls due to simulations exiting early, or external race conditions such as from shared filesystems.

I think slipping in these extra barriers with minimal documentation doesn't break anything, but preempts some possibly confusing situations.

**Comment by peterkasson** [comment]
I think we want to be very careful with such an idea.  This breaks a lot of performance benefits.  It's a Bad Thing to have barriers as a general case.

**Comment by peterkasson** [comment]
PS a barrier before the simulations start might be reasonable, but we'd want to examine the idea carefully.

**Comment by eirrgang** [comment]
Yes, actually, we avoid some of the problems this is intended to solve by just using a sub communicator for the simulations. In other cases, I now recall that we had talked about a more idiomatic synchronization through special WorkElements.

This whole issue is very different in gmxapi 0.0.7 and gmxapi 0.1 that I'm not sure what, if anything, needs to be done in 0.0.7. Maybe it *is* just a documentation issue.
